[Workers] QR code revision (#2173)
cloudflare/cloudflare-docs: products/workers/src/content/tutorials/build-a-qr-code-generator/index.md

## Overview

In this tutorial, you will build and publish a serverless function that generates QR codes, using Cloudflare Workers.

This guide will teach you how to build and publish serverless functions. This guide does not assume prior experience with serverless functions or Cloudflare Workers.

> If you would like to skip straight to the code, the final version of the codebase is [available on GitHub](https://github.com/signalnerve/workers-qr-code-generator). You can take code provided in the example repository, customize it, and deploy it for use in your own projects.

Reviewer: ```suggestion
If you would like to skip straight to the code, the final version of the codebase is [available on GitHub](https://github.com/signalnerve/workers-qr-code-generator). You can take the code provided in the example repository, customize it, and deploy it for use in your own projects.
```

## Generate

> Cloudflare’s command-line tool for managing Worker projects, Wrangler, supports various templates — pre-built collections of code that make it easy to get started writing Workers. You will make use of the default JavaScript template to start building your project.

Reviewer: Can we link to the doc page on Wrangler?

$ cd qr-code-generator
```

Wrangler templates are Git repositories. If you want to create your own templates, or use one from the [Template Gallery](/examples), there is a variety of options to help you get started.

Cloudflare’s `worker-template` includes support for building and deploying JavaScript-based projects. Inside of your new `qr-code-generator` directory, `index.js` represents the entry point to your Cloudflare Workers application.

> All Cloudflare Workers applications start by listening for `fetch` events. `fetch` events are triggered when a client makes a request to a Workers route. After a request is received by the Worker, your constructed response will return to the user. This tutorial will guide you through understanding how the request/response pattern works and how you can use it to build fully-featured applications.

Reviewer 2: ```suggestion
All Cloudflare Workers applications start by listening for `fetch` events, which are triggered when a client makes a request to a Workers route. After a request is received by the Worker, the response your application constructs will be returned to the user. This tutorial will guide you through understanding how the request/response pattern works and how you can use it to build fully-featured applications.
```

}
```

In your default `index.js` file, you can see that request/response pattern in action. The `handleRequest` constructs a new `Response` with the body text “Hello worker”, as well as an explicit `200` status code. 

> When a Worker receives a `fetch` event, the script must use `event.respondWith` to return the newly constructed response to the client. Your Cloudflare Worker script will serve new responses directly from [Cloudflare's edge network](https://www.cloudflare.com/network) instead of continuing to your origin server. A standard server would accept requests, and return responses. Cloudflare Workers allows you to respond quickly by constructing responses directly on the Cloudflare edge network.

Reviewer: ```suggestion
When a Worker receives a `fetch` event, the script must use `event.respondWith` to return the newly constructed response to the client. Your Cloudflare Worker script will serve new responses directly from [Cloudflare's edge network](https://www.cloudflare.com/network) instead of continuing to your origin server. A standard server would accept requests and return responses. Cloudflare Workers allows you to respond quickly by constructing responses directly on the Cloudflare edge network.
```

## Build

> Any project you publish to Cloudflare Workers can make use of modern Javascript tooling like ES modules, NPM packages, and [async/await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function) functions to build your application. In addition to writing serverless functions, you can use Workers to [build full applications](/tutorials/build-a-slackbot) using the same tooling and process as what you will be building in this tutorial.

Reviewer: ```suggestion
Any project you publish to Cloudflare Workers can make use of modern JavaScript tooling like ES modules, NPM packages, and [async/await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function) functions to build your application. In addition to writing serverless functions, you can use Workers to [build full applications](/tutorials/build-a-slackbot) using the same tooling and process as in this tutorial.
```

> The QR code generator you will build in this tutorial will be a serverless function that runs at a single route and receives requests. Given text sent inside of that request (such as URLs, or strings), the function will encode the text into a QR code, and serve the QR code as a PNG response.

Reviewer 2: ```suggestion
The QR code generator you will build in this tutorial will be a serverless function that runs on a single route and receives requests. Each request will contain a text message (a URL, for example), which the function will encode into a QR code. The function will then respond with the QR code in a PNG image format.
```

> The QR code generator you will build in this tutorial will be a serverless function that runs on a single route and receives requests. Each request will contain a text message (a URL, for example), which the function will encode into a QR code. The function will then respond with the QR code in a PNG image format.

Reviewer: Not sure if you need the "a" in "in [a] PNG image format."

}
```

> If an incoming request is not a `POST`, the function returns `undefined`. Since your focus is only about incoming `POST` requests, return a new `Response` with a [405 status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405) if the incoming request is not a `POST`:

Reviewer 2: ```suggestion
Currently, if an incoming request is not a `POST`, the function will return `undefined`. However, a Worker always needs to return a `Response`. Since the function should only accept incoming `POST` requests, return a new `Response` with a [`405` status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405) if the incoming request is not a `POST`:
```

}
```

> At this point, you have established the the basic flow of `handleRequest`. You will now set up a response to incoming valid requests. If a `POST` request comes in, the function should generate a QR code. To start, move the “Hello worker!” response into a new function, `generate`, which will ultimately contain the bulk of our function’s logic:

Reviewer: ```suggestion
At this point, you have established the basic flow of `handleRequest`. You will now set up a response to incoming valid requests. If a `POST` request comes in, the function should generate a QR code. To start, move the “Hello worker!” response into a new function, `generate`, which will ultimately contain the bulk of our function’s logic:
```

### Building a QR Code

> All projects deployed to Cloudflare Workers support NPM packages. This support makes it easy to rapidly build out functionality in your serverless functions. The [`qr-image`](https://github.com/alexeyten/qr-image) package is a great way to take text, and encode it into a QR code. The package supports generating the QR codes in a number of file formats (such as PNG, the default, and SVG), and configuring other aspects of the generated QR code. In the command line, install and save `qr-image` to your project’s `package.json`:

Reviewer: ```suggestion
All projects deployed to Cloudflare Workers support NPM packages. This support makes it easy to rapidly build out functionality in your serverless functions. The [`qr-image`](https://github.com/alexeyten/qr-image) package is a great way to take text and encode it into a QR code. The package supports generating the QR codes in a number of file formats (such as PNG, the default, and SVG), and configuring other aspects of the generated QR code. In the command line, install and save `qr-image` to your project’s `package.json`:
```

$ npm install --save qr-image
```

> To use the `qr-image` package, configure the `type` to `"webpack"`. This will tell Wrangler to use [Webpack](/cli-wrangler/webpack) to package your project for deployment. (Learn more about [`type` configuration](/cli-wrangler/configuration).)

Reviewer 2: ```suggestion
To use the `qr-image` package, configure the `type` to `"webpack"`. This instructs Wrangler to use [Webpack](/cli-wrangler/webpack) to package your project for deployment. (Learn more about [`type` configuration](/cli-wrangler/configuration).)
```

}
```

> By default, the QR code is generated as a PNG. Construct a new instance of `Response`, passing in the PNG data as the body, and a `Content-Type` header of `image/png`: this will allow browsers to properly parse the data coming back from your serverless function as an image:

Reviewer: ```suggestion
By default, the QR code is generated as a PNG. Construct a new instance of `Response`, passing in the PNG data as the body, and a `Content-Type` header of `image/png` — this will allow browsers to properly parse the data coming back from your serverless function as an image:
```
To avoid two colons in a row.

}
```

> ### Testing in an application UI

Author: @signalnerve same question as the previous tutorial about the usage of the acronym UI here :)

Reviewer 3: same thing! the UI referred to here is the one we build in the tutorial itself

> The serverless function will work if a user sends a `POST` request to a route, but it would be great to _also_ be able to test the function with a proper interface. At this point in the tutorial, if any request is received by your function that is not a `POST`, a `500` response is returned. The new version of `handleRequest` should return a new `Response` with a static HTML body, instead of the `500` error:

Author: @signalnerve I thought that written above it said if the request is NOT a POST request, a 405 error is returned? Line 85. Did something change?

Reviewer 3: you're right! that should be a 405, do you mind fixing that real quick? nice catch!!

> The serverless function will work if a user sends a `POST` request to a route, but it would be great to _also_ be able to test the function with a proper interface. At this point in the tutorial, if any request is received by your function that is not a `POST`, a `500` response is returned. The new version of `handleRequest` should return a new `Response` with a static HTML body, instead of the `500` error:

Author: ```suggestion
The serverless function will work if a user sends a `POST` request to a route, but it would be great to _also_ be able to test the function with a proper interface. At this point in the tutorial, if any request is received by your function that is not a `POST`, a `405` response is returned. The new version of `handleRequest` should return a new `Response` with a static HTML body, instead of the `405` error:
```

> The serverless function will work if a user sends a `POST` request to a route, but it would be great to _also_ be able to test the function with a proper interface. At this point in the tutorial, if any request is received by your function that is not a `POST`, a `405` response is returned. The new version of `handleRequest` should return a new `Response` with a static HTML body, instead of the `405` error:

Reviewer 2: ```suggestion
The serverless function will work if a user sends a `POST` request to a route, but it would be great to _also_ be able to test the function with a proper interface. At this point in the tutorial, if any request is received by your function that is not a `POST`, a `405` response is returned. The new version of `handleRequest` should return a new `Response` with a static HTML document instead of the `405` error:
```

> The serverless function will work if a user sends a `POST` request to a route, but it would be great to _also_ be able to test the function with a proper interface. At this point in the tutorial, if any request is received by your function that is not a `POST`, a `405` response is returned. The new version of `handleRequest` should return a new `Response` with a static HTML document instead of the `405` error:

Reviewer: Not sure if we should use italics in `_also_`.
